src/material/schematics/ng-generate/m3-theme/index.ts

```diff
@@ -192,3 +207,110 @@ export default function (options: Schema): Rule {
     createThemeFile(themeScss, tree, options.directory);
   };
 }
+
+/**
+ * The hue map produced by `material-color-utilities` may miss some neutral hues depending on
+ * the provided colors. This function estimates the missing hues based on the generated ones
+ * to ensure that we always produce a full palette. See #29157.
+ *
+ * This is a TypeScript port of the logic in `core/theming/_palettes.scss#_patch-missing-hues`.
+ */
+function patchMissingHues(
+  palettes: Map<string, Map<number, string>>,
+): Map<string, Map<number, string>> {
+  const neutral = palettes.get('neutral');
+
+  if (!neutral) {
+    return palettes;
+  }
+
+  let newNeutral: Map<number, string> | null = null;
+
+  for (const [hue, {prev, next}] of NEUTRAL_HUES) {
```

Comment:
The logic here can be simplified, but I wanted to avoid mutating the original map.